--- Loader/PhpFileLoader.php
+++ Loader/PhpFileLoader.php
@@ -61,7 +61,9 @@
             $callback = $load($path, $this->env);

             if (\is_object($callback) && \is_callable($callback)) {
-                $this->executeCallback($callback, new ContainerConfigurator($this->container, $this, $this->instanceof, $path, $resource, $this->env), $path);
+                $reflectionFunction = new \ReflectionFunction($callback);
+                $containerConfiguratorClass = $reflectionFunction->getParameters()[0]->getType()->getName();
+                $this->executeCallback($callback, new $containerConfiguratorClass($this->container, $this, $this->instanceof, $path, $resource, $this->env), $path);
             }
         } finally {
             $this->instanceof = [];
@@ -112,31 +114,32 @@
             }
             $type = $reflectionType->getName();

-            switch ($type) {
-                case ContainerConfigurator::class:
-                    $arguments[] = $containerConfigurator;
-                    break;
-                case ContainerBuilder::class:
-                    $arguments[] = $this->container;
-                    break;
-                case FileLoader::class:
-                case self::class:
-                    $arguments[] = $this;
-                    break;
-                case 'string':
-                    if (null !== $this->env && 'env' === $parameter->getName()) {
-                        $arguments[] = $this->env;
+            if (is_a($type, ContainerConfigurator::class, true)) {
+                $arguments[] = $containerConfigurator;
+            } else {
+                switch ($type) {
+                    case ContainerBuilder::class:
+                        $arguments[] = $this->container;
                         break;
-                    }
-                    // no break
-                default:
-                    try {
-                        $configBuilder = $this->configBuilder($type);
-                    } catch (InvalidArgumentException|\LogicException $e) {
-                        throw new \InvalidArgumentException(sprintf('Could not resolve argument "%s" for "%s".', $type.' $'.$parameter->getName(), $path), 0, $e);
-                    }
-                    $configBuilders[] = $configBuilder;
-                    $arguments[] = $configBuilder;
+                    case FileLoader::class:
+                    case self::class:
+                        $arguments[] = $this;
+                        break;
+                    case 'string':
+                        if (null !== $this->env && 'env' === $parameter->getName()) {
+                            $arguments[] = $this->env;
+                            break;
+                        }
+                        // no break
+                    default:
+                        try {
+                            $configBuilder = $this->configBuilder($type);
+                        } catch (InvalidArgumentException|\LogicException $e) {
+                            throw new \InvalidArgumentException(sprintf('Could not resolve argument "%s" for "%s".', $type . ' $' . $parameter->getName(), $path), 0, $e);
+                        }
+                        $configBuilders[] = $configBuilder;
+                        $arguments[] = $configBuilder;
+                }
             }
         }

